### ISSUE TITLE
Enhancement: [switch-exhaustiveness-check] require default cases for non-union types

### ISSUE THREAD

**Issue by ST-DDT** [typescript-eslint/typescript-eslint#7862]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/switch-exhaustiveness-check/

### Description

I propose that the `switch-exhaustiveness-check` checks `switch` statements that operate on non-union-types and require them to have a default case to ensure they are exhaustive.

### Fail

```typescript
const value: number = Math.floor(Math.random() * 3);
switch (value) {
  case 0:
    console.log("zero");
    return;
  case 1:
    console.log("one");
    return;
}
```

### Pass

```typescript
const value: number = Math.floor(Math.random() * 3);
switch (value) {
  case 0:
    console.log("zero");
    return;
  case 1:
    console.log("one");
    return;
  default:
    console.log("higher");
    return;
  // Or
  default:
    // Do nothing
}
```

### Additional Info

Since it would break projects that have the `switch-exhaustiveness-check` enabled, but don't have all their switch statements fully exhaustive, this should probably locked behind a default false option. Maybe this could be turned on by default with the next major version.

This option can not be replaced with [`default-case`](https://eslint.org/docs/latest/rules/default-case) due to #7539.

- https://github.com/typescript-eslint/typescript-eslint/pull/7539

Originally searched for during:

- https://github.com/faker-js/faker/pull/2519

**Comment by ST-DDT** [comment]
Duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/2959
